### ISSUE TITLE
sjoin leaves field values undefined when there are duplicate field names 

### ISSUE THREAD

**Issue by brahn** [chrispahm/spatialmerge#2]
I encountered the same problem as #1: when the left and right objects share field names, the result of `sjoin` leaves right-side field values undefined.  

For a test case, use:
* [Sacramento Neighborhoods](https://data.cityofsacramento.org/datasets/49f20f1612ae4f0a9292eb65f8bd4013_0/about) for the left-side shapes
* [Sacramento Parks Public Restrooms](https://data.cityofsacramento.org/datasets/b9e7fa6d1d104833b3f04268d7f682dc_0/about) for the right-side points.

```javascript
import { readFileSync } from 'fs';
import { join } from 'path';
import { sjoin } from 'spatialmerge';

const readJson = (filepath: string) => JSON.parse(readFileSync(filepath, 'utf8'));
const neighborhoods = readJson(join(__dirname, 'data/Neighborhoods.geojson'));
const restrooms = readJson(join(__dirname, 'data/Parks_Public_Restrooms.geojson'));
const mergeResult = sjoin(restrooms, neighborhoods, { how: "inner", op: "intersects" });

console.log(mergeResult.features[0].properties)
// console output:
{
  OBJECTID: 216,
  Facility: 'AIRPORT LITTLE LEAGUE PARK',
  JURISDICTION: 'City',
  WithinPark: 'Yes',
  GlobalID: '{75922410-F154-4E4A-8315-E418D5D7E6E6}',
  index_right: undefined,
  NAME: undefined,
  SHAPE_Length: undefined,
  SHAPE_Area: undefined
}
```
I would expect the `undefined` fields to be populated with values from the properties of the matching neighborhood:
```javascript
console.log(neighborhoods.features[110].properties)
// console output:
{
  OBJECTID: 111, 
  NAME: 'Golf Course Terrace', 
  SHAPE_Length: 0.08183311421559303, 
  SHAPE_Area: 0.0001623340769874396
}
```
 
I believe the problem may be a typo on line 24 of [helpers.js](https://github.com/chrispahm/spatialmerge/blob/54e985054dae2824318e51f0cc6715311353236b/src/helpers.js#L20-L27
). Should that line use the value `object2[key]` rather than `object1[key]`?
https://github.com/chrispahm/spatialmerge/blob/54e985054dae2824318e51f0cc6715311353236b/src/helpers.js#L20-L27

-----

I started putting together a pull request for a fix but then I encountered some additional issues, which gave me enough pause that I thought I should stop and check my understanding.  The additional issues:

* The `lsuffix` and `rsuffix` arguments are not populated in [this call to `mergeWith`](https://github.com/brahn/spatialmerge/blob/dff5f13693a9bfa58741f25e891b4a4d03868e4f/src/sjoin.js#L116-L122), which results in fields being overwritten rather than disambiguated: https://github.com/chrispahm/spatialmerge/blob/dff5f13693a9bfa58741f25e891b4a4d03868e4f/src/sjoin.js#L116-L122

* [This calculation of `duplicates`](https://github.com/brahn/spatialmerge/blob/dff5f13693a9bfa58741f25e891b4a4d03868e4f/src/helpers.js#L12) doesn't seem right:
https://github.com/chrispahm/spatialmerge/blob/dff5f13693a9bfa58741f25e891b4a4d03868e4f/src/helpers.js#L12
For example:
    ```javascript
    const allKeys = ['a', 'b', 'c', 'c', 'a'];
    const uniq = [...new Set(allKeys)];
    const duplicates = allKeys.filter(key => uniq.indexOf(key) === -1);
    console.log(duplicates);
    // actual output: []
    // expected output: ['a', 'c']  (or some permutation)
    ```

**Comment by brahn** [comment]
Here's a draft PR: https://github.com/chrispahm/spatialmerge/pull/3  Recommend reading one commit at a time.

With this change, the first feature in the test case described above is as follows:
```javascript
console.log(mergeResult.features[0].properties)
// console output:
{
  Facility: 'AIRPORT LITTLE LEAGUE PARK',
  GlobalID: '{75922410-F154-4E4A-8315-E418D5D7E6E6}',
  index_right: 110,
  JURISDICTION: 'City',
  NAME: 'Golf Course Terrace',
  OBJECTID_left: 216,
  OBJECTID_right: 111,
  SHAPE_Area: 0.0001623340769874396,
  SHAPE_Length: 0.08183311421559303,
  WithinPark: 'Yes'
}
```

Note that the second commit causes the `sjoin` test to fail, and the following new properties to appear in the first feature: 
```
  ADMIN: 'Egypt',
  ISO_A2: 'EG'
```

**Comment by chrispahm** [comment]
Hey @brahn,

Thanks a lot for opening an issue + PR and taking the time for the code review!

The property joining section is really weak, I was planning on optimizing / refactoring but then forgot about it 😅 so thanks a lot for working on that!

I'm currently on vacation so will only be able to have a look next week. Just wanted to let you know that I appreciate the work 😊

**Comment by brahn** [comment]
Thank you!

**Comment by brahn** [comment]
Hi @chrispahm just a friendly check-in :-)  Thanks!

**Comment by chrispahm** [comment]
Sorry! Will do the review tomorrow 😊

**Comment by chrispahm** [comment]
Alright, I went through the code (see #3) and it looks good to me!

A final todo would be fixing the tests, which were however already failing before the changes from #3. In principle, I'd prefer to check the merge / sjoin results against Geopandas to ensure everything works as expected (see Jupyter Notebook in the test dir). I started to work on that, but it seems Geopandas changes the index positions within the feature collection in the output file so I have to correct that first.

This is unfortunately more work than I envisioned, so I'll continue to work on it on Monday!

**Comment by brahn** [comment]
Sounds like a good path forward, thank you for the update!

**Comment by chrispahm** [comment]
Hey @brahn,
I just finished both fixing the existing tests and adding your test case 
https://github.com/chrispahm/spatialmerge/blob/593192cee8c537f03a95fa654dcac444b5af4163/test/test.js#L140

to #3, and it seems to work ok! 
Could I ask you to have a look and do a final review round before this gets merged? Thanks!

**Comment by brahn** [comment]
Looks great!  Thanks for tackling the test cases :-)

**Comment by chrispahm** [comment]
Awesome, just merged #3 and released as v1.0.1 🎉 
Thanks again for your PR and your commitment! 

Next up would probably be some performance checks and optimizations, but that's a different issue...

**Comment by brahn** [comment]
🎉 🎆
